test/distributed/fsdp/test_fsdp_exec_order.py

```diff
@@ -0,0 +1,145 @@
+        params_to_freeze = self.layer2.parameters() if self.use_alt_path \
+            else self.layer1.parameters()
+        params_to_unfreeze = self.layer1.parameters() if self.use_alt_path \
+            else self.layer2.parameters()
+        for param in params_to_freeze:
+            param.requires_grad = False
+        for param in params_to_unfreeze:
+            param.requires_grad = True
+        self.use_alt_path = not self.use_alt_path
+
+    @staticmethod
+    def wrap(sharding_strategy: ShardingStrategy, device: torch.device):
+        model = Model()
+        model.layer1 = FSDP(model.layer1, sharding_strategy=sharding_strategy)
+        model.layer2 = FSDP(model.layer2, sharding_strategy=sharding_strategy)
+        fsdp_model = FSDP(model, sharding_strategy=sharding_strategy)
+        return fsdp_model.to(device)
+
+
+class TestFSDPExecOrder(FSDPTest):
+    @property
+    def device(self):
+        return torch.device("cuda")
+
+    @skip_if_lt_x_gpu(2)
+    @parametrize(
+        "sharding_strategy",
+        [ShardingStrategy.FULL_SHARD, ShardingStrategy.SHARD_GRAD_OP],
+    )
+    def test_invalid_first_iter_fwd_order(self, sharding_strategy: ShardingStrategy):
+        """Tests that FSDP errors if the forward order differs across ranks in
+        the first iteration."""
+        # Rank 0 runs the forward pass in one order and all other ranks run in
+        # different order
+        fsdp_model = Model.wrap(sharding_strategy, self.device)
+        if self.rank != 0:
+            fsdp_model.flip_path()
+        inp = fsdp_model.module.get_input(self.device)
+        error_regex = "^(Forward pass execution order differs across ranks)"
+        with self.assertRaisesRegex(RuntimeError, error_regex):
+            fsdp_model(*inp)
+
+    @skip_if_lt_x_gpu(2)
+    @parametrize(
+        "sharding_strategy",
+        [ShardingStrategy.FULL_SHARD, ShardingStrategy.SHARD_GRAD_OP],
+    )
+    @parametrize("iters_before_path_change", [1, 3])
+    def test_invalid_later_iter_fwd_order(
+        self,
+        sharding_strategy: ShardingStrategy,
+        iters_before_path_change: int,
+    ):
+        """Tests that FSDP warns the user if the forward order changes after
+        the first iteration."""
+        # On the first iteration, all ranks run the same order, and on the next
+        # iteration, all but rank 0 run in a different order
+        fsdp_model = Model.wrap(sharding_strategy, self.device)
+        for _ in range(iters_before_path_change):
+            inp = fsdp_model.module.get_input(self.device)
+            output = fsdp_model(*inp)
+            loss = fsdp_model.module.get_loss(inp, output).to(self.device)
+            fsdp_model.module.run_backward(loss)
+        context = self.assertWarns(expected_warning=UserWarning) \
+            if self.rank != 0 else suppress()
+        if self.rank != 0:
+            fsdp_model.flip_path()
+        inp = fsdp_model.module.get_input(self.device)
+        with context:
```

Comment:
Yes, this will be a false positive. However, since the warning message is written to say that a change has been detected, not necessarily that something is wrong, hopefully it is not too confusing for the user.